pytorch_lightning/core/memory.py

```diff
@@ -1,164 +1,243 @@
-        self.out_sizes = out_sizes
-        assert len(in_sizes) == len(out_sizes)
-
-    def get_layer_names(self) -> None:
-        """ Collect Layer Names """
-        mods = self.named_modules()
-        names = []
-        layers = []
-        for name, m in mods:
-            names += [name]
-            layers += [str(m.__class__)]
-
-        layer_types = [x.split('.')[-1][:-2] for x in layers]
-
-        self.layer_names = names
-        self.layer_types = layer_types
-
-    def get_parameter_sizes(self) -> None:
-        """ Get sizes of all parameters in `model`. """
-        mods = self.named_modules()
-        sizes = []
-        for _, m in mods:
-            p = list(m.parameters())
-            modsz = [np.array(param.size()) for param in p]
-            sizes.append(modsz)
-
-        self.param_sizes = sizes
-
-    def get_parameter_nums(self) -> None:
-        """ Get number of parameters in each layer. """
-        param_nums = []
-        for mod in self.param_sizes:
-            all_params = 0
-            for p in mod:
-                all_params += np.prod(p)
-            param_nums.append(all_params)
-        self.param_nums = param_nums
-
-    def make_summary(self) -> None:
+    def __str__(self):
         """
         Makes a summary listing with:
 
-        Layer Name, Layer Type, Input Size, Output Size, Number of Parameters
+        Layer Name, Layer Type, Number of Parameters, Input Sizes, Output Sizes
         """
-        arrays = [['Name', self.layer_names],
-                  ['Type', self.layer_types],
-                  ['Params', list(map(get_human_readable_count, self.param_nums))]]
-        if self.model.example_input_array is not None:
-            arrays.append(['In sizes', self.in_sizes])
-            arrays.append(['Out sizes', self.out_sizes])
+        arrays = [
+            [" ", list(map(str, range(len(self._layer_summary))))],
+            ["Name", self.layer_names],
+            ["Type", self.layer_types],
+            ["Params", list(map(get_human_readable_count, self.param_nums))],
+        ]
+        if self._model.example_input_array is not None:
+            arrays.append(["In sizes", self.in_sizes])
+            arrays.append(["Out sizes", self.out_sizes])
+
+        return _format_summary_table(*arrays)
+
+    def __repr__(self):
+        return str(self)
 
-        self.summary = _format_summary_table(*arrays)
 
-    def summarize(self) -> None:
-        self.get_layer_names()
-        self.get_parameter_sizes()
-        self.get_parameter_nums()
+def parse_batch_shape(batch: Any) -> Union[str, List]:
+    if hasattr(batch, "shape"):
+        return list(batch.shape)
 
-        if self.model.example_input_array is not None:
-            self.get_variable_sizes()
-        self.make_summary()
+    if isinstance(batch, (list, tuple)):
+        shape = [parse_batch_shape(el) for el in batch]
+        return shape
+
+    return UNKNOWN_SIZE
 
 
 def _format_summary_table(*cols) -> str:
```

Comment:
we already have a table printer for Profiler, so lets general that function and use it here too

Reply:
there is no utility for table printing in Profiler that can be extracted. 
the shape, columns and rows of the table are all hardcoded and differ between Simple and AdvancedProfiler. 

not sure how to extract this, seems this is a larger effort that does not fit well into this pr here.

Reply:
https://github.com/PyTorchLightning/pytorch-lightning/blob/d6646e151a5c8c1ccfa4c4b3e6c859b0f81a1166/pytorch_lightning/profiler/profilers.py#L138-L151

Reply:
yes that's exactly what I was looking at.